module/zfs/vdev.c

```diff
@@ -1157,22 +1156,20 @@ vdev_open_children(vdev_t *vd)
 	 * in a single thread so that the same thread holds the
 	 * spa_namespace_lock
 	 */
-	if (vdev_uses_zvols(vd)) {
+	if (vdev_uses_zvols(vd) || taskq_member(system_taskq, curthread)) {
 		for (c = 0; c < children; c++) {
 			vd->vdev_child[c]->vdev_open_error =
 			    vdev_open(vd->vdev_child[c]);
 			vd->vdev_nonrot &= vd->vdev_child[c]->vdev_nonrot;
 		}
 		return;
 	}
-	tq = taskq_create("vdev_open", children, minclsyspri,
-	    children, children, TASKQ_PREPOPULATE);
 
 	for (c = 0; c < children; c++)
-		VERIFY(taskq_dispatch(tq, vdev_open_child, vd->vdev_child[c],
-		    TQ_SLEEP) != 0);
+		VERIFY(taskq_dispatch(system_taskq, vdev_open_child,
+		    vd->vdev_child[c], TQ_SLEEP) != 0);
 
-	taskq_destroy(tq);
+	taskq_wait_outstanding(system_taskq, 0);
```

Comment:
It will wait for all pending tasks which have be dispatched to the system_taskq but not yet processed.  That might involve other unrelated tasks so you're right it's not ideal.  This is almost certainly why existing code creates it's own taskq for this purpose.  Anyway, I'll see about making it slightly more clever.